libgalois/src/PropertyGraph.cpp

```diff
@@ -551,3 +551,124 @@ katana::SortNodesByDegree(katana::PropertyGraph* pg) {
+      new katana::LargeArray<uint64_t>();
+  katana::LargeArray<uint32_t>* out_dests_symmetric =
+      new katana::LargeArray<uint32_t>();
+  katana::LargeArray<uint64_t> out_indices_symmetric_temp;
+
+  out_indices_symmetric->allocateInterleaved(topology.num_nodes());
+  out_indices_symmetric_temp.allocateInterleaved(topology.num_nodes());
+
+  auto edge_range = topology.edge_range(0);
+  (*out_indices_symmetric)[0] = edge_range.first;
+  katana::do_all(
+      katana::iterate(uint64_t{1}, topology.num_nodes()), [&](uint64_t n) {
+        auto edge_range = topology.edge_range(n);
+        (*out_indices_symmetric)[n] = (edge_range.second - edge_range.first);
+      });
+
+  katana::do_all(
+      katana::iterate(uint64_t{0}, topology.num_nodes()),
+      [&](uint64_t n) {
+        auto edge_range = topology.edge_range(n);
+        // update the out_indices for the symmetric topology
+        for (auto e = edge_range.first; e != edge_range.second; ++e) {
+          auto dest = out_dests_view[e];
+          //Ignore self edges
+          if (n != dest) {
+            __sync_fetch_and_add(&((*out_indices_symmetric)[dest]), 1);
+          }
+        }
+      },
+      katana::steal());
+
+  // Compute prefix sum
+  for (uint64_t n = 1; n < topology.num_nodes(); ++n) {
+    (*out_indices_symmetric)[n] += (*out_indices_symmetric)[n - 1];
+  }
+
+  uint64_t num_nodes_symmetric = topology.num_nodes();
+  uint64_t num_edges_symmetric =
+      (*out_indices_symmetric)[num_nodes_symmetric - 1];
+
+  // Temp LargeArray for computing new destination positions
+  out_indices_symmetric_temp[0] = 0;
+  katana::do_all(
+      katana::iterate(uint64_t{1}, topology.num_nodes()),
+      [&](uint64_t n) {
+        out_indices_symmetric_temp[n] = (*out_indices_symmetric)[n - 1];
+      },
+      katana::no_stats());
+
+  out_dests_symmetric->allocateInterleaved(num_edges_symmetric);
+
+  // Update graph topology with the original edges + reverse edges
+  katana::do_all(
+      katana::iterate(uint64_t{0}, topology.num_nodes()),
+      [&](uint64_t src) {
+        auto edge_range = topology.edge_range(src);
+        // e = start index into edge array for a particular node
+        uint64_t e = edge_range.first;
+
+        // get all outgoing edges of a particular node in the
+        // and add reverse edges.
+        while (e < edge_range.second) {
+          // destination node
+          auto dst = out_dests_view[e];
+
+          // Add original edge
+          auto e_new_src =
+              __sync_fetch_and_add(&(out_indices_symmetric_temp[src]), 1);
+          (*out_dests_symmetric)[e_new_src] = dst;
+
+          if (dst != src) {
+            // Add reverse edge
+            auto e_new_dst =
+                __sync_fetch_and_add(&(out_indices_symmetric_temp[dst]), 1);
+            (*out_dests_symmetric)[e_new_dst] = src;
+            // TODO (gill) copy edge data to "new" array
+            // edgeDataCopy(edgeData_new, edgeData, e_new, e);
```

Comment:
(1) This is required for original edges too.

(2) There is a design choice: (i) duplicate edge properties or (ii) use pointers to original properties. I'm in favor of (ii) but that would some changes to PropertyGraph. I think similar changes would anyway be required in PropertyGraph v0.2.

Reply:
I am just focusing on the topology for now as that is what I need for most of the analytics algorithms. Once we move to v0.2 we will re-evaluate this.